### PR TITLE
Solution backend update

### PR DESCRIPTION
**Pull request by joshepvodka**
## About the PR
Changes and adds a couple methods to the ``Solution`` class.
The maximum and minimum temperatures of a solution are now equal to the boiling and melting points of the reagents it contains.
Exemple: If a beaker contains water you cannot heat it past 100C
Extra heat is stored as BoilingLatentHeat for adding boiling reactions in a different PR in the future.
Same thing happens for missing heat if you try to chill the solution past the highest melting point it has.

This would be part of the now closed #21823, I'm splitting it up into smaller PRs

## Why

To add stuff that makes chemistry interesting.

## Technical details
Added BoilingLatentHeat and MeltingLatentHeat variables on Solution.cs for accurately calculating phase transition reactions.
Added LowestBoilingPointReagent and HighestMeltingPointReagent variables on Solution.cs to cache what reagents limit the highest and lowest temperature the solution can be respectively.
Added BoilingLatentHeat and MeltingLatentHeat fields to ReagentPrototype to accurately calculating phase transition reactions.
Also changed 1 or 2 lines in many files because i changed a couple variable names.
I think the code is very well commented for reviewers to figure out what exactly is happening.

There's different fields for boiling and melting latent heats in the case someone pours liquid nitrogen into water, making the water freeze at the same time the nitrogen boils. And other reactions like that to be added in the hopefully not so far future.
Everything still works the same.

I removed the Solution_Tests.cs file because they'll never work and I'll never find a way to make them work without help.

## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

### CODE REVIEW

**Comment by Partmedia** [comment]
Thanks for your PR.

Latent heat preserves temperature because the extra heat is going into vaporizing the liquid. Since we don't have reagent phase changes yet, this extra heat seems like it would just be going nowhere. You *can* actually superheat water if you bring the pressure high enough.

Would you be willing to think about how gas vaporization might work? If not, let's see what we can do about this.

**Comment by joshepvodka** [comment]
> Thanks for your PR.
> 
> Latent heat preserves temperature because the extra heat is going into vaporizing the liquid. Since we don't have reagent phase changes yet, this extra heat seems like it would just be going nowhere. You _can_ actually superheat water if you bring the pressure high enough.
> 
> Would you be willing to think about how gas vaporization might work? If not, let's see what we can do about this.

I have already actualy, on PR #21823. I closed that one because it was too big, but I already had working code for vaporization of water. Also I'm working on a new chem-atmos PR that will have a boiling water example of the reaction properly working.
I know about the superheating water thing or having ice at room temperature but IMO i don't think those are important enough to put into the game.

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (Partmedia) [inline Content.Shared/Chemistry/Components/Solution.cs:107]: Could you please explain why this API breakage is needed?

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (joshepvodka) [reply Content.Shared/Chemistry/Components/Solution.cs:491]: To make the AddReagent code less spaghetti-y and to have less overloaded AddReagent functions. Also i just had to add a PrototypeManager arg on every function because it is needed to get the lowest and highest phase transition temperature reagents.

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (joshepvodka) [reply Content.Shared/Chemistry/Components/Solution.cs:107]: I think the code looks better if heat capacity is just a variable with private set for HeatCapacity. Also I didn't think _heatCapacityDirty was needed, so no need to check for it when you want to get the heat capacity.

**Comment by joshepvodka** [comment]
> Please see the inline comments. My thoughts are:
> 
>     * This change needlessly breaks the API and as result touches too many unrelated systems that happen to use solutions.
> 
>     * This could be a good candidate to merge by itself if the breakage is reduced.
> 
>     * This could also be a good candidate to merge together with liquid to gas conversion, since then the phase change energy wouldn't be going nowhere.

Why is API breakage so bad? As far as i know, everything still works just fine, and API didn't get any messier.
Also, I think liquid gas conversions are better suited for a different PR, as it brings many changes to ReactionPrototypes, ReagentPrototypes and the ChemicalReactionSystem.

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (metalgearsloth) [inline Content.Shared/Chemistry/Components/Solution.cs:104]: If you're going to start caching this then a test is required.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (joshepvodka) [reply Content.Shared/Chemistry/Components/Solution.cs:104]: the ``TestTemperatureCalculations`` test on ``Content.IntegrationTests\Tests\Chemistry\SolutionSystemTests.cs`` doesn't do that already?

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (ElectroJr) [reply Content.Shared/Chemistry/Components/Solution.cs:491]: Resolving the prototype manager and updating all the properties is expensive and will significantly slow down adding reagents, it should defer the processing as much as possible. PrototypeManager should also be a mandatory argument wherever possible to avoid people not passing it in

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (ElectroJr) [reply Content.Shared/Chemistry/Components/Solution.cs:107]: the idea is to intentionally defer updating the heat capacity until its required to make modifying solutions faster. This is also why the cached heat capacity wasn't publicly available - so that people didn't use it without first checking that the heat capacity is up to date

File: Content.Tests/Shared/Chemistry/SolutionTests.cs
Comment (ElectroJr) [inline Content.Tests/Shared/Chemistry/SolutionTests.cs:1]: Why were all of these tests removed?

Comment (joshepvodka) [reply Content.Tests/Shared/Chemistry/SolutionTests.cs:1]: Because solutions now need to get prototype info on boiling points and such, and it kept throwing "prototypes not loaded" errors. Also, I have no idea on how to make these tests work. I've asked on #howdoicode but got no answers. If someone helps me out I'll gladly put them back.

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (joshepvodka) [reply Content.Shared/Chemistry/Components/Solution.cs:107]: using protoman to get a prototype, even if we don't resolve it, is really expensive then?

**Comment by joshepvodka** [comment]
Changes:
Never gonna Resolve the PrototypeManager.
I'm caching frequently accessed ReagentPrototype proprieties like SpecificHeat and BoilingPoint so i only need to protoMan.Index once when a new reagent is added to a solution.
Changed some code on SplitSolution to make it work better with those changes.

File: Content.Shared/Chemistry/Components/Solution.cs
Comment (ElectroJr) [reply Content.Shared/Chemistry/Components/Solution.cs:107]: it is comparatively expensive for something that needs to happen very frequently. Solutions are used by a lot of other component/systems and there are many things that modify them every tick.

**Comment by ElectroJr** [comment]
Note that this PR will conflict pretty heavily with #21916.

In general, I think it would be cleaner if solutions were first refactored to replace all reagent prototype ID strings with a direct reference to the actual prototype, and then adding boiling/freezing after that. Then you wouldn't have to pass around the prototype manager at all, nor would you need the separate cache. Though it would still be important to defer calculations until needed. I'll bring up replacing the id strings in the next maintainer meeting.

Also, sorry if it feels like were are being very nitpicky or the PR takes a long time. Having things like boiling points and other advanced reagent properties is a feature we want, but solutions are important and affect many systems, so it is important that the implementation is both designed well / nice to use,  and that it doesn't hurt performance. Things like lungs, metabolism, stomachs, bloodstream, etc all use reagents, and those systems have previously had performance issues whenever there were many players.

**Comment by joshepvodka** [comment]
That's completely fine, this PR should be as robust as possible for merging, take as long as you need.

After #21916 gets merged I can work on adapting my code to fit it, most of the logic shouldn't change anyways.
>I'll bring up replacing the id strings in the next maintainer meeting.

Please give me an update after that happens, a comment under this PR is fine, I check my github often.

**Comment by metalgearsloth** [comment]
Other PR got updated.

**Comment by AJCM-git** [comment]
@joshepvodka are you coming back to this?

**Comment by joshepvodka** [comment]
Not coming back, someone else can take over if they want.

**Pull request closed without merging**
